torch_geometric/explain/base.py

```diff
@@ -0,0 +1,121 @@
+from abc import abstractmethod
+from typing import Optional
+
+import torch
+
+from torch_geometric.data import Data
+from torch_geometric.explain.explanations import Explanation
+
+
+class ExplainerAlgorithm(torch.nn.Module):
+    """Abstract class for explanation algorithms."""
+    @abstractmethod
+    def loss(self, y_hat: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+        r"""
+        This method compute the loss to be used for the explanation algorithm.
+        Subclasses should override this method to define their own loss.
+
+        Args:
+            y_hat (torch.Tensor): the output of the explanation algorithm.
+                (e.g. the forward pass of the model with the mask applied).
+            y (torch.Tensor): the reference output.
+        """
+
+    @torch.no_grad()
+    def get_initial_prediction(
+        self,
+        model: torch.nn.Module,
+        g: Data,
+        batch: Optional[torch.Tensor] = None,
+        **kwargs,
+    ):
+        """Return the initial prediction of the model.
+
+        Args:
+            model (torch.nn.Module): the model to explain.
+            g (torch_geometric.data.Data): the input graph.
+            batch (torch.Tensor, optional): batch indicator.
+        """
+        out = model(
+            g,
+            **dict(batch=batch, **kwargs),
+        )
+        return out
+
+    @abstractmethod
+    def forward(
+        self,
+        x: torch.Tensor,
+        edge_index: torch.Tensor,
+        *,
+        model: torch.nn.Module,
+        target: torch.Tensor,
+        target_index: int = 0,
+        batch: Optional[torch.Tensor] = None,
+        task_level: str = "graph",
+        return_type: str = "regression",
```

Comment:
Sorry naive question: Is it likely for a single explainer to support multiple return types?

Reply:
If you look at the gradient based methods, they would support both classification and regression tasks? But this might be added in the `supports` method of the `ExplainerAlgorithm`.

Reply:
Makes sense. I haven't worked with many different approaches to this so I wasn't sure. Thanks